### PR TITLE
Local exception cleanup

### PR DESCRIPTION
**Pull request by bernardnormier**
This is an initial C++ exception cleanup:

 - local exception are no longer implemented with a helper
 - removed ice_clone
 - updated Python to use exception_ptr

Only partially done: remove ice_throw

In modern C++, you're not supposed to clone exception or ask them to throw themselves. The correct pattern is to hold an exception_ptr and rethrow it as needed.

### CODE REVIEW

File: python/modules/IcePy/Util.cpp [hunk 1]
```diff
@@ -954,8 +958,39 @@ IcePy::convertException(const Ice::Exception& ex)
...
+        ostr << e.what();
+        string str = ostr.str();
+
+        type = lookupType("Ice.UnknownException");
+        assert(type);
+        p = createExceptionInstance(type);
+        if (p.get())
+        {
+            PyObjectHandle s = createString(str);
+            PyObject_SetAttrString(p.get(), STRCAST("unknown"), s.get());
```
Comment (pepone) [inline python/modules/IcePy/Util.cpp:986]: Cant we just pass the `const char*` returned by `e.what()` into createString?

```
PyObjectHandle s = createString(e.what());
```

**Review by pepone — APPROVED** [review]

File: cpp/include/IceBT/Types.h [hunk 2]
```diff
@@ -26,29 +17,24 @@
 #    endif
 #endif
 
+#if defined(__clang__)
+#    pragma clang diagnostic push
+#    pragma clang diagnostic ignored "-Wshadow-field-in-constructor"
```
Comment (externl) [inline cpp/include/IceBT/Types.h:22]: Is this just for `reason`? Can we just rename the field to `_reason`?

File: cpp/include/IceBox/Service.h [hunk 3]
```diff
@@ -70,28 +57,29 @@ namespace IceBox
          * @param line The line number at which the exception was raised, typically __LINE__.
          * @param reason The reason for the failure.
          */
-        FailureException(const char* file, int line, const ::std::string& reason)
-            : ::Ice::LocalExceptionHelper<FailureException, ::Ice::LocalException>(file, line),
-              reason(reason)
+        FailureException(const char* file, int line, std::string reason) noexcept
+            : LocalException(file, line),
+              reason(std::move(reason))
```
Comment (externl) [inline cpp/include/IceBox/Service.h:62]: Why does this one generate a constructor shadow warning?

File: cpp/include/IceUtil/Exception.h [hunk 4]
```diff
@@ -21,16 +21,16 @@ namespace IceUtil
     {
     public:
         /**
-         * Constructs the exception. Equivalent to Exception(nullptr, 0).
+         * Default constructor. Equivalent to Exception(nullptr, 0).
          */
-        Exception();
+        Exception() noexcept;
```
Comment (externl) [inline cpp/include/IceUtil/Exception.h:26]: Do we ever use this one?

File: cpp/src/Ice/OutgoingAsync.cpp [hunk 5]
```diff
@@ -740,24 +740,23 @@ OutgoingAsync::response()
...
                 _is.read(operation, false);
 
-                unique_ptr<RequestFailedException> ex;
                 switch (replyStatus)
                 {
                     case replyObjectNotExist:
                     {
-                        ex.reset(new ObjectNotExistException(__FILE__, __LINE__));
+                        throw ObjectNotExistException{__FILE__, __LINE__, ident, facet, operation};
                         break;
```
[the remark this answers is not in the document]
Comment (externl) [reply cpp/src/Ice/OutgoingAsync.cpp:748]: I'd keep it anyway.

**Review by externl — APPROVED** [review]
Looks good!

File: cpp/test/Ice/background/TestI.cpp [hunk 6]
```diff
@@ -83,7 +84,7 @@ BackgroundControllerI::readReady(bool enable, const Current&)
 void
 BackgroundControllerI::readException(bool enable, const Current&)
 {
-    _configuration->readException(enable ? new SocketException(__FILE__, __LINE__) : 0);
+    _configuration->readException(enable ? make_exception_ptr(SocketException(__FILE__, __LINE__)) : exception_ptr{});
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply cpp/test/Ice/background/TestI.cpp:87]: Yes, you're right.
https://en.cppreference.com/w/cpp/error/exception_ptr meets the requirements of NullablePointer so we can use nullptr.

File: python/modules/IcePy/Util.cpp [hunk 7]
```diff
@@ -954,8 +958,39 @@ IcePy::convertException(const Ice::Exception& ex)
...
+        }
+    }
+    catch (...)
+    {
+        ostr << "unknown c++ exception";
+        string str = ostr.str();
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply python/modules/IcePy/Util.cpp:992]: Fixed.

File: python/modules/IcePy/Util.cpp (hunk 1 above)
Comment (bernardnormier) [reply python/modules/IcePy/Util.cpp:986]: Yes, createString takes a string_view. Fixed.

File: cpp/include/IceUtil/Exception.h (hunk 4 above)
Comment (bernardnormier) [reply cpp/include/IceUtil/Exception.h:26]: Yes, we use the default constructor quite a bit. For example, user exceptions usually do use this ctor.

File: cpp/include/Ice/LocalException.h [hunk 8]
```diff
@@ -1122,96 +813,66 @@ namespace Ice
...
+        std::string str;
     };
 
     /**
      * This exception is raised if an illegal identity is encountered.
      * \headerfile Ice/Ice.h
      */
-    class ICE_CLASS(ICE_API) IllegalIdentityException
-        : public LocalExceptionHelper<IllegalIdentityException, LocalException>
+    class ICE_API IllegalIdentityException : public LocalException
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply cpp/include/Ice/LocalException.h:850]: Yes, it would be good to prune some local exceptions. They are too numerous.

**Pull request merged**